### ISSUE TITLE
Use of `A | B as c` fails to format correctly 

### ISSUE THREAD

**Issue by dsyme** [fsprojects/fantomas#2289]
Found this while formatting the F# compiler. It's not a blocker, see workaround below

Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAdATugdgFwE8AHRMAFTAF4wBBMAHzACEtcAbRPMAMzAEMIlVjggQAtnzzQAFvzAB3AJZ5pw0QzAAKOoyb8AzmABGigDRg2fI4jYBGAJRgAtAD4160beGRT4KPs48RRwAc31UMABtPwhQDygAEkISCMhgvB8Y0UgEgDc%2BNgBXZDRo%2BIgABl9yqABJHCR8AGVFAC9kavKAFiyIAF0sgF9O2N7E5JKodMz4nPyiybLy2xGPSABZPgAPABlgxB3EUJUZ5YAmCrGB9WGsuLWk4knII-gAeW49nEQmwg5T9RzArFCJLeJnVaAgCiDQ%2BXw6YwgkDY3EgVyGq3ugMeKTQkCMsFgbAB2TywMWiLAAGZIdl6gFMHgAGLBAoAOUQ8jY%2BxJ6jwmGK6JumLGOQmqQJRN5iXmINKlK6tKRTSIfGgiCYiG4sEwiAACnxMHwxJxEJhpRB%2BYL4tdRLd1FjSeK8ZLibSgQtQZSAKxKqAqtUarU6g6weRm6B8AL1XKwSNBWA4C1gK2IIV2kWzHHPV3Sj1yqKUgBsfsgAfVmu1uoAqkQSJhI9GcLH44pE8nuAUAumIPbRI6kdmJYS3aKyZ75TUAOyl8tBquIADCln0%2BkXif0VugeB1Ha7aZtGLuY%2BdUFz7vHBbBHgAHLPVRXg7r1ogxNZzX7O2xu4fhces6e%2BIjnml4UjUACc96BpWIbrmw7afvuPZgH2owAU8w5ShespgcsVSUmWD6IDQ3B4Ga65iBIyapshqFgAO4wYS6wHYeSXo1LYKwEXOMG6k0r6KHA8FJoh34Hh4tq9pmDyAeeY44ex5xQeqJFkZg-FiIJRIIZSNG-hm-4yUxZ4sfJbGTssNLcURNCYLAhQNAAIjYiiaWp1ECuJ6iSSh0nYoB0ysROhYcYqBGbFstTcGQ0hHFCYlNNIOp4AA6oo8AnH6XSXPpUmGf5xlpPgIEKRZ8S2L64XbPU3CKFsbz1pIOpQlsRC6qubYiT6OUSUeDonoVgVmcF17qLYJZVVsABKiBwJg8BpRl0jJtltF%2BU6g3FUFV6UrYM6TTNc3wGyhRvmaHy1GRYjhH6thrflG24lAYiFGwQTct8TI6hIeBqWQGHbbh5V3gRh06vA6yve9%2BxfZgP3uaWMiGmqakAPpKEtaK5b5D2DgFW3DTtHGQZNtlGgQbyYHsm6LZlCo9d5fX9gNT1FRkgOKeC%2BE1Bs2xk3wFNU4om4nWdmAXVdN27fd-XoazL1vYoH2ILD8Nmv9KQc2VHhnFxPP84L1N4JDivK6rkgIwRSNGtuZro%2BlJwy8zcvPENWalSF5RnBCk0AGrkkwwTwMEIS08tfo3gzog%2BXRDFipt7OE0DOtWTzEVMg526dYHDQh2HK1R-0TNoUZrNuw8HujaIZxhWn2yObAeAAOKcC1bWIB1ib5363qF2AMfrXjhVye75me%2BClVp1DSv7Ew8HQAA1kwNsL5w%2BhvDgTTGkuRKnV1NRfj%2BvV-rLpc5qZo8jZSZwTTzHJcrPnDhkcGuIM5tU4MonU0A0L5i1LB8kLY1jizc%2BWEk6cx1vtfW3IQg4AzjgLOiYWhwMkIUXUZBYD1EaHgSQnU9xiSdiXAqrMR4VzHlXCAZwQYwLUjgPBuRQyhD-u%2Bd%2BwQv4bgIUfRmJ9nZn0wqOS%2BRMvYkynoreEOxjRGHgHwZcsB9Ah3vsrLhXlo7F3oqAgRJUKGUipNzcokA25mlckcXBbAADSiBEBEGwaY%2Boy8%2BCIPDpSQ%2Bqii68OIY9MBgjyFXxqFSPWBi56OIXvCfQtl7INA5JuRAENp7Kzbu1BRnC-R6WPgZU%2BJDvHaL8eUKkPseZMENLxN%2BwtoCYFcqyMi8Bqyf0TM5aAlgjQJn3uUVxRCNEuy0VrceHgqSpwMUYipJp8AFF%2BHZQom4BREF%2BAQf4oluFqI8Z0-heJy7Ykrro2uBjLHWIiqLd8Hxgk4FCfsABeE%2B4D1xoxUhF9fHCPiFSSeBjxmCWNggBEgDCHAL8H0ToaR9BMgUREdpgwgA)

#### Code

```fsharp

type T = A | B

let f a  =
    match a with
    | (A | B as bi, x) ->
        1

```
            

#### Result

```fsharp
type T =
    | A
    | B

let f a =
    match a with
    | (A
       | B as bi,
       label1) -> 1

```
            
#### Problem description

Please describe here the Fantomas problem you encountered.
                    Check out our [Contribution Guidelines](https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md#bug-reports).

### Workaround

The workaround is to parenthesize, e.g.
```fsharp

type T = A | B

let f a  =
    match a with
    | ((A | B) as bi, x) ->
        1
```

#### Extra information

- [ ] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas master branch at 2022-05-30T18:11:41Z - 5e927521516445116087af12c76d83e0dec01332

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#ignore-files-fantomasignore)?</sub>

**Comment by nojaf** [comment]
Hello, this rings a bell.
If I recall correctly, the main problem is that we format each `SynPat.Or` inside a clause slightly differently using [ASTContext](https://github.com/fsprojects/fantomas/blob/5e927521516445116087af12c76d83e0dec01332/src/Fantomas.Core/CodePrinter.fs#L30-L31).
`ASTContext` is evil, however:
https://github.com/fsprojects/fantomas/blob/5e927521516445116087af12c76d83e0dec01332/src/Fantomas.Core/CodePrinter.fs#L16-L18

I think this problem could be solved by removing that record field and trying to format `SynPat.Or` differently when encountered in the pattern of a clause handler.

Are you interested in submitting a PR for this?

**Comment by nojaf** [comment]
I believe this problem is addressed due to the refactor of `ASTContext` in a7f5b3ea71cad3a4efa4aface2edef8da4b88c91.
